## Porto Seguro’s Safe Driver Prediction

### Data Preparation & Exploration

    Visual inspection of your data
    Defining the metadata
    Descriptive statistics
    Handling imbalanced classes
    Data quality checks
    Exploratory data visualization
    Feature engineering
    Feature selection
    Feature scaling
   
   
Porto Seguro는 브라질의 자동차 보험 회사입니다. 본 경진 대회의 목적은 어떤 차주가 내년에 보험 청구를 할 확률을 예측하는 겁니다.

<br>데이터는 59만 개의 훈련 데이터와 89만 개의 테스트 데이터로 구성되어 있습니다. 테스트 데이터가 훈련 데이터보다 많습니다. 그리고 Null 값이 np.NaN이 아닌 -1로 되어있습니다. 또한, Feature가 무엇을 뜻하는지 제시하지 않았다는 것이 특징입니다. 보통은 특정 Feature가 무엇을 나타내는지 알려줍니다. 예를 들어 집 값을 예측하는 데이터에서 Feature 중 방 크기(Area)가 있습니다. 방 크기가 크면 집 값이 올라갈 것이라는 예상을 해볼 수 있습니다. 하지만 Porto Seguro 컴피티션에는 각 Feature가 무엇을 의미하는지 알 수 없습니다. 데이터를 안내한 부분에도 나와있지 않고, Feature의 이름으로 유추하기도 어렵습니다. 다만, Feature가 binary인지, categorical인지, oridnal인지, nominal인지만 구분할 수 있을 뿐입니다. 보안상 공개를 하지 않는 것 같습니다.

<br>target은 보험 청구를 한다(=1), 보험 청구를 하지 않는다(=0)인 binary 데이터입니다. 0과 1로 구성되어 있으며 0이 1보다 압도적으로 많습니다. 

In [142]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler

from sklearn.feature_selection import VarianceThreshold
from sklearn.feature_selection import SelectFromModel
from sklearn.utils import shuffle
from sklearn.ensemble import RandomForestClassifier

pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 100)

In [143]:
# DEBUG = True

# if DEBUG:
#     NROWS = 10000
# else:
#     NROWS = None

In [144]:
#%%time
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

In [145]:
# 데이터 싸이즈가 큰 경우 EDA 할 떄는 샘플링을 해서 진행한다. 
# train = pd.read_csv('train.csv', nrows = None)
# test = pd.read_csv('test.csv',  nrows = None)
# train = train.sample(frac=0.2)

In [146]:
train.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0              0              0              1              0              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              0              1              0              0              0   
4              0              1              0              0              0   

   ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  ps_ind_14  \
0              0              0              0              0          0   
1              0              0              0              0          0   
2              0              0              0              0          0   
3              0              0              0              0          0   
4              0              0              0              0          0   

   ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_reg_01  \
0         11              0              1              0        0.7   
1          3              0              0              1        0.8   
2         12              1              0              0        0.0   
3          8              1              0              0        0.9   
4          9              1              0              0        0.7   

   ps_reg_02  ps_reg_03  ps_car_01_cat  ps_car_02_cat  ps_car_03_cat  \
0        0.2   0.718070             10              1             -1   
1        0.4   0.766078             11              1             -1   
2        0.0  -1.000000              7              1             -1   
3        0.2   0.580948              7              1              0   
4        0.6   0.840759             11              1             -1   

   ps_car_04_cat  ps_car_05_cat  ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  \
0              0              1              4              1              0   
1              0             -1             11              1              1   
2              0             -1             14              1              1   
3              0              1             11              1              1   
4              0             -1             14              1              1   

   ps_car_09_cat  ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  \
0              0              1             12          2   0.400000   
1              2              1             19          3   0.316228   
2              2              1             60          1   0.316228   
3              3              1            104          1   0.374166   
4              2              1             82          3   0.316070   

   ps_car_13  ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  \
0   0.883679   0.370810   3.605551         0.6         0.5         0.2   
1   0.618817   0.388716   2.449490         0.3         0.1         0.3   
2   0.641586   0.347275   3.316625         0.5         0.7         0.1   
3   0.542949   0.294958   2.000000         0.6         0.9         0.1   
4   0.565832   0.365103   2.000000         0.4         0.6         0.0   

   ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  \
0           3           1          10           1          10           1   
1           2           1           9           5           8           1   
2           2           2           9           1           8           2   
3           2           4           7           1           8           4   
4           2           2           6   

In [147]:
train.tail()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
595207  1488013       0          3              1         10              0   
595208  1488016       0          5              1          3              0   
595209  1488017       0          1              1         10              0   
595210  1488021       0          5              2          3              1   
595211  1488027       0          0              1          8              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              1              0              0   
595210              0              0              0              1   
595211              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
595207              1              0              0              0   
595208              1              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
595207              0          0         13              1              0   
595208              0          0          6              1              0   
595209              0          0         12              1              0   
595210              0          0         12              1              0   
595211              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
595207              0        0.5        0.3   0.692820             10   
595208              0        0.9        0.7   1.382027              9   
595209              0        0.9        0.2   0.659071              7   
595210              0        0.9        0.4   0.698212             11   
595211              0        0.1        0.2  -1.000000              7   

        ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  ps_car_05_cat  \
595207              1             -1              0              1   
595208              1             -1              0             -1   
595209              1             -1              0             -1   
595210              1             -1              0             -1   
595211              0             -1              0             -1   

        ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
595207              1              1              1              0   
595208             15              0              0              2   
595209              1              1              1              2   
595210             11              1              1              2   
595211              0              1              0              2   

        ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  \
595207              1             31          3   0.374166   0.684631   
595208              1             63          2   0.387298   0.972145   
595209              1             31          3   0.397492   0.596373   
595210              1            101          3   0.374166   0.764434   
595211              1             34          2   0.400000   0.932649   

        ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  \
595207   0.385487   2.645751         0.4         0.5         0.3           3   
595208  -1.000000   3.605551         0.2         0.2         0.0           2   
595209   0.398748   1.732051         0.4         0.0         0.3           3   
595210   0.384968   3.162278         0.0         0.7         0.0           4   
595211   0.378021   3.741657         0.4         0.0         0.5           2   

        ps_calc_05  ps_calc_06  

In [148]:
train.shape

(595212, 59)

In [149]:
test.shape

(892816, 58)

In [150]:
cat_cols = [col for col in train.columns if 'cat' in col]

In [151]:
cat_cols

['ps_ind_02_cat',
 'ps_ind_04_cat',
 'ps_ind_05_cat',
 'ps_car_01_cat',
 'ps_car_02_cat',
 'ps_car_03_cat',
 'ps_car_04_cat',
 'ps_car_05_cat',
 'ps_car_06_cat',
 'ps_car_07_cat',
 'ps_car_08_cat',
 'ps_car_09_cat',
 'ps_car_10_cat',
 'ps_car_11_cat']

In [152]:
train[cat_cols[0]].value_counts()

 1    431859
 2    123573
 3     28186
 4     11378
-1       216
Name: ps_ind_02_cat, dtype: int64

In [153]:
#카테고리별  unique() 갯수 표시

for col in cat_cols:
    print(col, train[col].nunique())

ps_ind_02_cat 5
ps_ind_04_cat 3
ps_ind_05_cat 8
ps_car_01_cat 13
ps_car_02_cat 3
ps_car_03_cat 3
ps_car_04_cat 10
ps_car_05_cat 3
ps_car_06_cat 18
ps_car_07_cat 3
ps_car_08_cat 2
ps_car_09_cat 6
ps_car_10_cat 3
ps_car_11_cat 104


In [154]:
print("train set과 test set에서 cloumns차이:", set(train.columns) - set(test.columns))

train set과 test set에서 cloumns차이: {'target'}


In [155]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 595212 entries, 0 to 595211
Data columns (total 59 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              595212 non-null  int64  
 1   target          595212 non-null  int64  
 2   ps_ind_01       595212 non-null  int64  
 3   ps_ind_02_cat   595212 non-null  int64  
 4   ps_ind_03       595212 non-null  int64  
 5   ps_ind_04_cat   595212 non-null  int64  
 6   ps_ind_05_cat   595212 non-null  int64  
 7   ps_ind_06_bin   595212 non-null  int64  
 8   ps_ind_07_bin   595212 non-null  int64  
 9   ps_ind_08_bin   595212 non-null  int64  
 10  ps_ind_09_bin   595212 non-null  int64  
 11  ps_ind_10_bin   595212 non-null  int64  
 12  ps_ind_11_bin   595212 non-null  int64  
 13  ps_ind_12_bin   595212 non-null  int64  
 14  ps_ind_13_bin   595212 non-null  int64  
 15  ps_ind_14       595212 non-null  int64  
 16  ps_ind_15       595212 non-null  int64  
 17  ps_ind_16_

### meta data

Concretely we will store:

    role: input, ID, target
    level: nominal, interval, ordinal, binary
    keep: True or False
    dtype: int, float, str

In [156]:
data = []
for f in train.columns:
    #role 선언
    if f == 'target':
        role ='target'
    elif f == 'id':
        role = 'id'
    else:
        role = 'input'
    # level 선언
    if 'bin' in f or f == 'target':
        level = 'binary'
    elif 'cat' in f or f == 'id':
        level = 'normal'
    elif train[f].dtype == float:
        level = 'interval'
    else:
        level = 'ordinal'
        
    #Id 를 제외한 다른 변수에 True 를 초기화
    keep = True
    if f =='id':
        keep = False
    
    #dtype정의
    dtype = train[f].dtype
    
    #메타변수 딕셔너리 만들기
    f_dict = {
        'varname': f,
        'role':role,
        'level':level,
        'keep':keep,
        'dtype': dtype
    }
    data.append(f_dict)

In [157]:
meta = pd.DataFrame(data,)

In [158]:
meta = pd.DataFrame(data,columns = ['varname','role','level','keep','dtype'])

In [159]:
meta.set_index('varname', inplace = True)

In [160]:
meta

role     level   keep    dtype
varname                                         
id                  id    normal  False    int64
target          target    binary   True    int64
ps_ind_01        input   ordinal   True    int64
ps_ind_02_cat    input    normal   True    int64
ps_ind_03        input   ordinal   True    int64
ps_ind_04_cat    input    normal   True    int64
ps_ind_05_cat    input    normal   True    int64
ps_ind_06_bin    input    binary   True    int64
ps_ind_07_bin    input    binary   True    int64
ps_ind_08_bin    input    binary   True    int64
ps_ind_09_bin    input    binary   True    int64
ps_ind_10_bin    input    binary   True    int64
ps_ind_11_bin    input    binary   True    int64
ps_ind_12_bin    input    binary   True    int64
ps_ind_13_bin    input    binary   True    int64
ps_ind_14        input   ordinal   True    int64
ps_ind_15        input   ordinal   True    int64
ps_ind_16_bin    input    binary   True    int64
ps_ind_17_bin    input    binary   True    int64
ps_ind_18_bin    input    binary   True    int64
ps_reg_01        input  interval   True  float64
ps_reg_02        input  interval   True  float64
ps_reg_03        input  interval   True  float64
ps_car_01_cat    input    normal   True    int64
ps_car_02_cat    input    normal   True    int64
ps_car_03_cat    input    normal   True    int64
ps_car_04_cat    input    normal   True    int64
ps_car_05_cat    input    normal   True    int64
ps_car_06_cat    input    normal   True    int64
ps_car_07_cat    input    normal   True    int64
ps_car_08_cat    input    normal   True    int64
ps_car_09_cat    input    normal   True    int64
ps_car_10_cat    input    normal   True    int64
ps_car_11_cat    input    normal   True    int64
ps_car_11        input   ordinal   True    int64
ps_car_12        input  interval   True  float64
ps_car_13        input  interval   True  float64
ps_car_14        input  interval   True  float64
ps_car_15        input  interval   True  float64
ps_calc_01       input  interval   True  float64
ps_calc_02       input  interval   True  float64
ps_calc_03       input  interval   True  float64
ps_calc_04       input   ordinal   True    int64
ps_calc_05       input   ordinal   True    int64
ps_calc_06       input   ordinal   True    int64
ps_calc_07       input   ordinal   True    int64
ps_calc_08       input   ordinal   True    int64
ps_calc_09       input   ordinal   True    int64
ps_calc_10       input   ordinal   True    int64
ps_calc_11       input   ordinal   True    int64
ps_calc_12       input   ordinal   True    int64
ps_calc_13       input   ordinal   True    int64
ps_calc_14       input   ordinal   True    int64
ps_calc_15_bin   input    binary   True    int64
ps_calc_16_bin   input    binary   True    int64
ps_calc_17_bin   input    binary   True    int64
ps_calc_18_bin   input    binary   True    int64
ps_calc_19_bin   input    binary   True    int64
ps_calc_20_bin   input    binary   True    int64

* Example to extract all normal variable that are not dropped

In [161]:
meta.loc[(meta.level =='normal') & (meta.keep)].index

Index(['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_car_01_cat',
       'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat',
       'ps_car_06_cat', 'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat',
       'ps_car_10_cat', 'ps_car_11_cat'],
      dtype='object', name='varname')

* Below the number of variables per role and level are displayed.

In [162]:
pd.DataFrame({'count': meta.groupby(['role','level'])['role'].size()}).reset_index()

role     level  count
0      id    normal      1
1   input    binary     17
2   input  interval     10
3   input    normal     14
4   input   ordinal     16
5  target    binary      1

In [163]:
meta.groupby(['role','level'])['role'].size()

role    level   
id      normal       1
input   binary      17
        interval    10
        normal      14
        ordinal     16
target  binary       1
Name: role, dtype: int64

* Descriptive statistics

We can also apply the describe method on the dataframe. However, it doesn't make much sense to calculate the mean, std, ... on categorical variables and the id variable. We'll explore the categorical variables visually later.

Thanks to our meta file we can easily select the variables on which we want to compute the descriptive statistics. To keep things clear, we'll do this per data type.

In [164]:
catagorical_feats = [col for col in train.columns if 'cat' in col]

In [165]:
v = meta[(meta.level == 'interval') & (meta.keep)].index
train[v].describe()

ps_reg_01      ps_reg_02      ps_reg_03      ps_car_12  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.610991       0.439184       0.551102       0.379945   
std         0.287643       0.404264       0.793506       0.058327   
min         0.000000       0.000000      -1.000000      -1.000000   
25%         0.400000       0.200000       0.525000       0.316228   
50%         0.700000       0.300000       0.720677       0.374166   
75%         0.900000       0.600000       1.000000       0.400000   
max         0.900000       1.800000       4.037945       1.264911   

           ps_car_13      ps_car_14      ps_car_15     ps_calc_01  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.813265       0.276256       3.065899       0.449756   
std         0.224588       0.357154       0.731366       0.287198   
min         0.250619      -1.000000       0.000000       0.000000   
25%         0.670867       0.333167       2.828427       0.200000   
50%         0.765811       0.368782       3.316625       0.500000   
75%         0.906190       0.396485       3.605551       0.700000   
max         3.720626       0.636396       3.741657       0.900000   

          ps_calc_02     ps_calc_03  
count  595212.000000  595212.000000  
mean        0.449589       0.449849  
std         0.286893       0.287153  
min         0.000000       0.000000  
25%         0.200000       0.200000  
50%         0.400000       0.500000  
75%         0.700000       0.700000  
max         0.900000       0.900000

* reg variables

only ps_reg_03 has missing values <br>
the range (min to max) differs between the variables. We could apply scaling (e.g. StandardScaler), but it depends on the classifier we will want to use.<br>


* car variables

ps_car_12 and ps_car_15 have missing values<br>
again, the range differs and we could apply scaling.<br>

* calc variables

no missing values<br>
this seems to be some kind of ratio as the maximum is 0.9<br>
all three _calc variables have very similar distributions<br>
<p>Overall, we can see that the range of the interval variables is rather small. Perhaps some transformation (e.g. log) is already applied in order to anonymize the data?</p>

In [166]:
v = meta[(meta.level == 'ordinal') & (meta.keep)].index
train[v].describe()

ps_ind_01      ps_ind_03      ps_ind_14      ps_ind_15  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        1.900378       4.423318       0.012451       7.299922   
std         1.983789       2.699902       0.127545       3.546042   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       2.000000       0.000000       5.000000   
50%         1.000000       4.000000       0.000000       7.000000   
75%         3.000000       6.000000       0.000000      10.000000   
max         7.000000      11.000000       4.000000      13.000000   

           ps_car_11     ps_calc_04     ps_calc_05     ps_calc_06  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        2.346072       2.372081       1.885886       7.689445   
std         0.832548       1.117219       1.134927       1.334312   
min        -1.000000       0.000000       0.000000       0.000000   
25%         2.000000       2.000000       1.000000       7.000000   
50%         3.000000       2.000000       2.000000       8.000000   
75%         3.000000       3.000000       3.000000       9.000000   
max         3.000000       5.000000       6.000000      10.000000   

          ps_calc_07     ps_calc_08     ps_calc_09     ps_calc_10  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        3.005823       9.225904       2.339034       8.433590   
std         1.414564       1.459672       1.246949       2.904597   
min         0.000000       2.000000       0.000000       0.000000   
25%         2.000000       8.000000       1.000000       6.000000   
50%         3.000000       9.000000       2.000000       8.000000   
75%         4.000000      10.000000       3.000000      10.000000   
max         9.000000      12.000000       7.000000      25.000000   

          ps_calc_11     ps_calc_12     ps_calc_13     ps_calc_14  
count  595212.000000  595212.000000  595212.000000  595212.000000  
mean        5.441382       1.441918       2.872288       7.539026  
std         2.332871       1.202963       1.694887       2.746652  
min         0.000000       0.000000       0.000000       0.000000  
25%         4.000000       1.000000       2.000000       6.000000  
50%         5.000000       1.000000       3.000000       7.000000  
75%         7.000000       2.000000       4.000000       9.000000  
max        19.000000      10.000000      13.000000      23.000000

* Binary variables

In [167]:
v = meta[(meta.level == 'binary') & (meta.keep)].index
train[v].describe()

target  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.036448       0.393742       0.257033       0.163921   
std         0.187401       0.488579       0.436998       0.370205   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       1.000000       1.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.185304       0.000373       0.001692       0.009439   
std         0.388544       0.019309       0.041097       0.096693   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_ind_13_bin  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.000948       0.660823       0.121081       0.153446   
std         0.030768       0.473430       0.326222       0.360417   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         0.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_calc_15_bin  ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  \
count   595212.000000   595212.000000   595212.000000   595212.000000   
mean         0.122427        0.627840        0.554182        0.287182   
std          0.327779        0.483381        0.497056        0.452447   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          0.000000        1.000000        1.000000        0.000000   
75%          0.000000        1.000000        1.000000        1.000000   
max          1.000000        1.000000        1.000000        1.000000   

       ps_calc_19_bin  ps_calc_20_bin  
count   595212.000000   595212.000000  
mean         0.349024        0.153318  
std          0.476662        0.360295  
min          0.000000        0.000000  
25%          0.000000        0.000000  
50%          0.000000        0.000000  
75%          1.000000        0.000000  
max          1.000000        1.000000

* A priori in the train data is 3.645%, which is strongly imbalanced.
* From the means we can conclude that for most variables the value is zero in most cases.

* Handling imbalanced classes

As we mentioned above the proportion of records with target=1 is far less than target=0. This can lead to a model that has great accuracy but does have any added value in practice. Two possible strategies to deal with this problem are:

* oversampling records with target=1
* undersampling records with target=0

There are many more strategies of course and MachineLearningMastery.com gives a nice overview. As we have a rather large training set, we can go for undersampling.

In [168]:
desired_apriori = 0.10

In [169]:
idx_0 = train[train.target == 0].index
idx_1 = train[train.target == 1].index

In [170]:
nb_0 = len(train.loc[idx_0])
nb_1 = len(train.loc[idx_1])

In [171]:
# Calculate the undersampling rate and resulting number of records with target=0


In [172]:
undersampling_rate = ((1-desired_apriori)*nb_1) / (nb_0*desired_apriori)

In [173]:
undersampling_rate

0.34043569687437886

In [174]:
undersampled_nb_0 = int(undersampling_rate*nb_0)
print('Rate to undersample records with target = 0: {}' .format(undersampling_rate))
print('Number of records with target=0 after undersampling: {}' .format(undersampled_nb_0))

Rate to undersample records with target = 0: 0.34043569687437886
Number of records with target=0 after undersampling: 195246


In [175]:
undersampled_idx = shuffle(idx_0, random_state = 37, n_samples = undersampled_nb_0)

In [176]:
idx_list = list(undersampled_idx)  +list(idx_1)

In [177]:
train.loc[idx_list].reset_index(drop = True)

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0        353762       0          1              1          1              1   
1        845629       0          5              1          7              1   
2       1449185       0          4              1          3              1   
3        834066       0          6              1          6              0   
4        436189       0          5              1          5              0   
...         ...     ...        ...            ...        ...            ...   
216935  1487473       1          5              1          7              0   
216936  1487566       1          1              1          5              0   
216937  1487716       1          0              1          6              1   
216938  1487748       1          0              2          2              0   
216939  1487866       1          1              2          1              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
0                   0              1              0              0   
1                   1              1              0              0   
2                   0              1              0              0   
3                   0              1              0              0   
4                   0              0              0              1   
...               ...            ...            ...            ...   
216935              3              0              0              1   
216936              0              0              0              1   
216937              0              0              1              0   
216938              0              1              0              0   
216939              0              0              0              1   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
0                   0              0              0              0   
1                   0              0              0              0   
2                   0              0              0              0   
3                   0              0              0              0   
4                   0              0              0              0   
...               ...            ...            ...            ...   
216935              0              0              0              0   
216936              0              0              0              0   
216937              0              0              0              0   
216938              0              0              0              0   
216939              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
0                   0          0         11              0              1   
1                   0          0          5              0              0   
2                   0          0          4              1              0   
3                   0          0          8              1              0   
4                   0          0         10              1              0   
...               ...        ...        ...            ...            ...   
216935              0          0          8              0              1   
216936              0          0          4              0              0   
216937              0          0          9              0              1   
216938              0          0          2              0              0   
216939              0          0         13              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
0                   0        0.6        0.6   0.838153             11   
1                   1        0.9        0.6   0.728440              7   
2                   0        0.9        0.6   0.869267             11   
3                   0        0.6        1.5   1.705872              6   
4                   0        0.8        0.8   1.086566              5   
...               .

* Data Quailty Checks
* Checking missing values

In [178]:
vars_with_missing = []

for f in train.columns:
    missings = train[train[f] == -1][f].count()
    if missings > 0:
        vars_with_missing.append(f)
        missings_perc = missings/train.shape[0]
        
        print('Variable {} has {} records ({:.2%}) with missing values'.format(f, missings, missings_perc))
        
print('In total, there are {} variables with missing values'.format(len(vars_with_missing)))

Variable ps_ind_02_cat has 216 records (0.04%) with missing values
Variable ps_ind_04_cat has 83 records (0.01%) with missing values
Variable ps_ind_05_cat has 5809 records (0.98%) with missing values
Variable ps_reg_03 has 107772 records (18.11%) with missing values
Variable ps_car_01_cat has 107 records (0.02%) with missing values
Variable ps_car_02_cat has 5 records (0.00%) with missing values
Variable ps_car_03_cat has 411231 records (69.09%) with missing values
Variable ps_car_05_cat has 266551 records (44.78%) with missing values
Variable ps_car_07_cat has 11489 records (1.93%) with missing values
Variable ps_car_09_cat has 569 records (0.10%) with missing values
Variable ps_car_11 has 5 records (0.00%) with missing values
Variable ps_car_12 has 1 records (0.00%) with missing values
Variable ps_car_14 has 42620 records (7.16%) with missing values
In total, there are 13 variables with missing values


* ps_car_03_cat and ps_car_05_cat have a large proportion of records with missing values. Remove these variables.
* For the other categorical variables with missing values, we can leave the missing value -1 as such.
* ps_reg_03 (continuous) has missing values for 18% of all records. Replace by the mean.
* ps_car_11 (ordinal) has only 5 records with misisng values. Replace by the mode.
* ps_car_12 (continuous) has only 1 records with missing value. Replace by the mean.
* ps_car_14 (continuous) has missing values for 7% of all records. Replace by the mean.

In [179]:
# Dropping the variables with too many missing values
vars_to_drop = ['ps_car_03_cat', 'ps_car_05_cat']
train.drop(vars_to_drop, inplace=True, axis=1)

In [180]:
meta.loc[(vars_to_drop),'keep'] = False 

In [181]:
# temp_series = train[['ps_car_01_cat', 'ps_car_02_cat', 'ps_reg_03']].groupby(['ps_car_01_cat', 'ps_car_02_cat'])

In [182]:
# temp_series.reset_index()

In [183]:
mean_imp = SimpleImputer(missing_values = -1, strategy = 'mean')
mode_imp = SimpleImputer(missing_values = -1, strategy = 'most_frequent' )

In [184]:
train['ps_reg_03'] = mean_imp.fit_transform(train[['ps_reg_03']]).ravel()
train['ps_car_12'] = mean_imp.fit_transform(train[['ps_car_12']]).ravel()
train['ps_car_14'] = mean_imp.fit_transform(train[['ps_car_14']]).ravel()
train['ps_car_11'] = mode_imp.fit_transform(train[['ps_car_11']]).ravel()

* Checking the cardinality of the categorical variables

In [294]:
v = meta[(meta.level == 'nominal') & (meta.keep)].index


In [ ]:

for f in v:
    dist_values = train[f].value_counts().shape[0]
    print('Variable {} has {} distinct values'.format(f, dist_values))

In [293]:
v

Index([], dtype='object', name='varname')

### 3강

In [ ]:
train_encoded, test_encoded = target_encode(train["ps_car_11_cat"], 
                             test["ps_car_11_cat"], 
                             target=train.target, 
                             min_samples_leaf=100,
                             smoothing=10,
                             noise_level=0.01)

In [223]:
trn_series=train["ps_car_11_cat"]
tst_series=test["ps_car_11_cat"]
# trn_series = pd.DataFrame(trn_series)
# tst_series = pd.DataFrame(tst_series)
target=train.target 
min_samples_leaf=100 
smoothing=10
noise_level=0.01

In [224]:
assert len(trn_series) == len(target)

In [225]:
assert trn_series.name == tst_series.name

In [226]:
temp = pd.concat([trn_series, target], axis =1)

In [228]:
temp.groupby(by = trn_series.name)[target.name].agg(['mean','count' ])

mean  count
ps_car_11_cat                 
1              0.040528   3331
2              0.037995   2553
3              0.051177   3185
4              0.060777   1596
5              0.041956  12513
...                 ...    ...
100            0.052418   4445
101            0.038289   7339
102            0.033761   2103
103            0.026667  24262
104            0.045015  85083

[104 rows x 2 columns]

In [229]:
# agg()방법 2
temp.groupby(by = trn_series.name)[target.name].mean()

ps_car_11_cat
1      0.040528
2      0.037995
3      0.051177
4      0.060777
5      0.041956
         ...   
100    0.052418
101    0.038289
102    0.033761
103    0.026667
104    0.045015
Name: target, Length: 104, dtype: float64

In [231]:
#agg()방법 3 함수

def max_min(x):
    return x.max() - x.min()

In [232]:
temp.groupby(by = trn_series.name)[target.name].agg(['mean','count' ,max_min])

mean  count  max_min
ps_car_11_cat                          
1              0.040528   3331        1
2              0.037995   2553        1
3              0.051177   3185        1
4              0.060777   1596        1
5              0.041956  12513        1
...                 ...    ...      ...
100            0.052418   4445        1
101            0.038289   7339        1
102            0.033761   2103        1
103            0.026667  24262        1
104            0.045015  85083        1

[104 rows x 3 columns]

In [235]:
averages = temp.groupby(by = trn_series.name)[target.name].agg(['mean','count'])

In [236]:
smoothing = 1 / (1+ np.exp(-(averages["count"]- min_samples_leaf)/smoothing))

In [237]:
prior = target.mean()

In [238]:
averages[target.name] = prior * (1-smoothing) + averages['mean']*smoothing

In [239]:
averages.drop(['mean','count'], axis = 1, inplace = True)

In [244]:
ft_trn_series = pd.merge(
        trn_series.to_frame(trn_series.name),
        averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on=trn_series.name,
        how='left')['average'].rename(trn_series.name + '_mean').fillna(prior)

In [249]:
ft_trn_series.index = trn_series.index

In [252]:
# test date에도 똑같이 
ft_tst_series = pd.merge(
        tst_series.to_frame(tst_series.name),
        averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on=tst_series.name,
        how='left')['average'].rename(trn_series.name + '_mean').fillna(prior)

In [253]:
ft_tst_series.index = tst_series.index

In [257]:
0.01* np.random.randn(ft_trn_series.shape[0])

array([ 0.00560042, -0.01845013,  0.00214498, ..., -0.00786963,
       -0.00792522,  0.00401739])

In [259]:
# Script by https://www.kaggle.com/ogrellier
# Code: https://www.kaggle.com/ogrellier/python-target-encoding-for-categorical-features
def add_noise(series, noise_level):
    return series * (1 + noise_level * np.random.randn(len(series)))

def target_encode(trn_series=None, 
                  tst_series=None, 
                  target=None, 
                  min_samples_leaf=1, 
                  smoothing=1,
                  noise_level=0):
    """
    Smoothing is computed like in the following paper by Daniele Micci-Barreca
    https://kaggle2.blob.core.windows.net/forum-message-attachments/225952/7441/high%20cardinality%20categoricals.pdf
    trn_series : training categorical feature as a pd.Series
    tst_series : test categorical feature as a pd.Series
    target : target data as a pd.Series
    min_samples_leaf (int) : minimum samples to take category average into account
    smoothing (int) : smoothing effect to balance categorical average vs prior  
    """ 
    assert len(trn_series) == len(target)
    assert trn_series.name == tst_series.name
    temp = pd.concat([trn_series, target], axis=1)
    # Compute target mean 
    averages = temp.groupby(by=trn_series.name)[target.name].agg(["mean", "count"])
    # Compute smoothing
    smoothing = 1 / (1 + np.exp(-(averages["count"] - min_samples_leaf) / smoothing))
    # Apply average function to all target data
    prior = target.mean()
    # The bigger the count the less full_avg is taken into account
    averages[target.name] = prior * (1 - smoothing) + averages["mean"] * smoothing
    averages.drop(["mean", "count"], axis=1, inplace=True)
    # Apply averages to trn and tst series
    ft_trn_series = pd.merge(
        trn_series.to_frame(trn_series.name),
        averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on=trn_series.name,
        how='left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    # pd.merge does not keep the index so restore it
    ft_trn_series.index = trn_series.index 
    ft_tst_series = pd.merge(
        tst_series.to_frame(tst_series.name),
        averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on=tst_series.name,
        how='left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    # pd.merge does not keep the index so restore it
    ft_tst_series.index = tst_series.index
    return add_noise(ft_trn_series, noise_level), add_noise(ft_tst_series, noise_level)

In [260]:
train_encoded, test_encoded = target_encode(train["ps_car_11_cat"], 
                             test["ps_car_11_cat"], 
                             target=train.target, 
                             min_samples_leaf=100,
                             smoothing=10,
                             noise_level=0.01)

In [261]:
train_encoded

0         0.038211
1         0.024188
2         0.031555
3         0.045650
4         0.026191
            ...   
595207    0.038863
595208    0.062441
595209    0.039164
595210    0.038198
595211    0.037154
Name: ps_car_11_cat_mean, Length: 595212, dtype: float64

In [262]:
train['ps_car_11_cat_te'] = train_encoded

In [263]:
train.drop('ps_car_11_cat', axis = 1, inplace = True)

In [265]:
meta.loc['ps_car_c11_cat', 'keep']  = False

In [266]:
test['ps_car_11_cat_te'] = test_encoded
test.drop('ps_car_11_cat', axis=1, inplace=True)

### EDA

In [286]:
v = meta[(meta.level == 'nominal')&(meta.keep)].index

In [287]:
sns.set(font_scale = 3)

In [289]:
for f in v:
    break
    plt.figure()
    fig, ax = plt.subplots(figsize=(20,10))
    # Calculate the percentage of target=1 per category value
    
    cat_perc = train[[f, 'target']].groupby([f], as_index = False)
    cat_perc.sort_values(by='target',ascending = False, inplace = true )
    
    sns.barplot(ax = ax, x = f, y = 'target', data = cat_perc, order = cat_perc[f])
    plt.ylabel('% target')
    plt.xlabel(f)
    plt.tick_params(axis='both', which='major')
    plt.show();